src/basic/macro.h

```diff
@@ -56,6 +60,29 @@
 #  endif
 #endif
 
+#if !defined(HAS_FEATURE_ADDRESS_SANITIZER)
+#  ifdef __SANITIZE_ADDRESS__
+#      define HAS_FEATURE_ADDRESS_SANITIZER 1
+#  elif defined(__has_feature)
+#    if __has_feature(address_sanitizer)
+#      define HAS_FEATURE_ADDRESS_SANITIZER 1
+#    endif
+#  endif
+#  if !defined(HAS_FEATURE_ADDRESS_SANITIZER)
+#    define HAS_FEATURE_ADDRESS_SANITIZER 0
+#  endif
+#endif
+
+/* Note: on GCC "no_sanitize_address" is a function attribute only, on llvm it may also be applied to global
+ * variables. We define a specific macro which knows this. Note that on GCC we don't need this decorator so much, since
+ * our primary usecase for this attribute is registration structures placed in named ELF sections which shall not be
+ * padded, but GCC doesn't pad those anyway if AddressSanitizer is enabled. */
```

Comment:
Turns out this isn't applicable to clang-3.4.2 or more precisely to
```
clang version 3.4.2 (tags/RELEASE_34/dot2-final)
Target: x86_64-redhat-linux-gnu
Thread model: posix
```
Here is what I've just got on CentOS:
```
[118/1733] Compiling C object 'src/libsystemd/src@libsystemd@@systemd_static@sta/sd-bus_bus-common-errors.c.o'.
FAILED: src/libsystemd/src@libsystemd@@systemd_static@sta/sd-bus_bus-common-errors.c.o
clang -Isrc/libsystemd/src@libsystemd@@systemd_static@sta -Isrc/libsystemd -I../src/libsystemd -Isrc/basic -I../src/basic -Isrc/shared -I../src/shared -Isrc/systemd -I../src/systemd -Isrc/journal -I../src/journal -Isrc/journal-remote -I../src/journal-remote -Isrc/nspawn -I../src/nspawn -Isrc/resolve -I../src/resolve -Isrc/timesync -I../src/timesync -I../src/time-wait-sync -Isrc/login -I../src/login -Isrc/udev -I../src/udev -Isrc/libudev -I../src/libudev -Isrc/core -I../src/core -I../src/libsystemd/sd-bus -I../src/libsystemd/sd-device -I../src/libsystemd/sd-event -I../src/libsystemd/sd-hwdb -I../src/libsystemd/sd-id128 -I../src/libsystemd/sd-netlink -I../src/libsystemd/sd-network -Isrc/libsystemd-network -I../src/libsystemd-network -I. -I../ -Xclang -fcolor-diagnostics -fsanitize=address -fno-omit-frame-pointer -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=gnu99 -g -Wextra -Werror=undef -Wmissing-include-dirs -Wold-style-definition -Wpointer-arith -Winit-self -Wfloat-equal -Werror=missing-prototypes -Werror=implicit-function-declaration -Werror=missing-declarations -Werror=return-type -Werror=incompatible-pointer-types -Werror=format=2 -Wstrict-prototypes -Wredundant-decls -Wmissing-noreturn -Wshadow -Wendif-labels -Wstrict-aliasing=2 -Wwrite-strings -Werror=overflow -Werror=shift-count-overflow -Wnested-externs -Wno-unused-parameter -Wno-missing-field-initializers -Wno-unused-result -Wno-error=nonnull -ffast-math -fno-common -fdiagnostics-show-option -fno-strict-aliasing -fvisibility=hidden -fstack-protector --param=ssp-buffer-size=4 -fPIE -Wno-typedef-redefinition -Wno-gnu-variable-sized-type-not-at-end -Werror=shadow -include config.h -g -O0 -ftrapv -fPIC -pthread -fvisibility=default  -MD -MQ 'src/libsystemd/src@libsystemd@@systemd_static@sta/sd-bus_bus-common-errors.c.o' -MF 'src/libsystemd/src@libsystemd@@systemd_static@sta/sd-bus_bus-common-errors.c.o.d' -o 'src/libsystemd/src@libsystemd@@systemd_static@sta/sd-bus_bus-common-errors.c.o' -c ../src/libsystemd/sd-bus/bus-common-errors.c
../src/libsystemd/sd-bus/bus-common-errors.c:10:1: error: '__no_sanitize_address__' attribute only applies to functions and methods
BUS_ERROR_MAP_ELF_REGISTER const sd_bus_error_map bus_common_errors[] = {
^
../src/libsystemd/sd-bus/bus-error.h:37:9: note: expanded from macro 'BUS_ERROR_MAP_ELF_REGISTER'
        _variable_no_sanitize_address_
        ^
../src/basic/macro.h:81:55: note: expanded from macro '_variable_no_sanitize_address_'
#define _variable_no_sanitize_address_ __attribute__((__no_sanitize_address__))
                                                      ^
1 error generated.
```
Though I'm not sure anyone is going to compile `systemd` with clang and asan there.

Reply:
Even without `ASan` it's failing to compile there:
```
[298/1733] Compiling C object 'src/shared/src@shared@@systemd-shared-239@sta/json.c.o'.
FAILED: src/shared/src@shared@@systemd-shared-239@sta/json.c.o
clang -Isrc/shared/src@shared@@systemd-shared-239@sta -Isrc/shared -I../src/shared -Isrc/basic -I../src/basic -Isrc/systemd -I../src/systemd -Isrc/journal -I../src/journal -Isrc/journal-remote -I../src/journal-remote -Isrc/nspawn -I../src/nspawn -Isrc/resolve -I../src/resolve -Isrc/timesync -I../src/timesync -I../src/time-wait-sync -Isrc/login -I../src/login -Isrc/udev -I../src/udev -Isrc/libudev -I../src/libudev -Isrc/core -I../src/core -I../src/libsystemd/sd-bus -I../src/libsystemd/sd-device -I../src/libsystemd/sd-event -I../src/libsystemd/sd-hwdb -I../src/libsystemd/sd-id128 -I../src/libsystemd/sd-netlink -I../src/libsystemd/sd-network -Isrc/libsystemd-network -I../src/libsystemd-network -I. -I../ -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/uuid -Xclang -fcolor-diagnostics -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=gnu99 -g -Wextra -Werror=undef -Wmissing-include-dirs -Wold-style-definition -Wpointer-arith -Winit-self -Wfloat-equal -Werror=missing-prototypes -Werror=implicit-function-declaration -Werror=missing-declarations -Werror=return-type -Werror=incompatible-pointer-types -Werror=format=2 -Wstrict-prototypes -Wredundant-decls -Wmissing-noreturn -Wshadow -Wendif-labels -Wstrict-aliasing=2 -Wwrite-strings -Werror=overflow -Werror=shift-count-overflow -Wnested-externs -Wno-unused-parameter -Wno-missing-field-initializers -Wno-unused-result -Wno-error=nonnull -ffast-math -fno-common -fdiagnostics-show-option -fno-strict-aliasing -fvisibility=hidden -fstack-protector --param=ssp-buffer-size=4 -fPIE -Wno-typedef-redefinition -Wno-gnu-variable-sized-type-not-at-end -Werror=shadow -include config.h -g -O0 -ftrapv -fPIC -pthread -fvisibility=default  -MD -MQ 'src/shared/src@shared@@systemd-shared-239@sta/json.c.o' -MF 'src/shared/src@shared@@systemd-shared-239@sta/json.c.o.d' -o 'src/shared/src@shared@@systemd-shared-239@sta/json.c.o' -c ../src/shared/json.c
In file included from ../src/shared/json.c:18:
../src/shared/json-internal.h:43:11: error: static_assert expression is not an integral constant expression
assert_cc((uintptr_t) _JSON_VARIANT_MAGIC_MAX < 4096U);
~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/basic/macro.h:343:23: note: expanded from macro 'assert_cc'
        static_assert(expr, #expr);
                      ^
../src/shared/json-internal.h:43:11: note: cast that performs the conversions of a reinterpret_cast is not allowed in a constant expression
../src/basic/macro.h:343:23: note: expanded from macro 'assert_cc'
        static_assert(expr, #expr);
                      ^
1 error generated.
```
Anyway I think it can be ignored for now.

Reply:
hmm, such an old clang even had asan? I am impressed...

Reply:
Interestingly, the official documentation at https://clang.llvm.org/docs/AddressSanitizer.html#conditional-compilation-with-has-feature-address-sanitizer doesn't say that the attribute can be applied to global variables so it seems that technically we're using an undocumented feature here :-)

Reply:
Oh never mind. I was apparently looking at the oldest documentation possible.